File: python/valid_anagram.py
# Solution 1.
# Time:b  
# Memory:

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        count_s, count_t = {}, {}
        
        for i in range(len(s)):
            count_s[s[i]] = count_s.get(s[i], 0) + 1
            count_t[t[i]] = count_t.get(t[i], 0) + 1
        for c in count_s:
            if count_s[c] != count_t.get(c, 0):
                return False
        return True

# Solution 2. Least Efficient
# Time:
# Memory:

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        
        count_s, count_t = {}, {}
        
        for i in range(len(s)):
            count_s[s[i]] = count_s.get(s[i], 0) + 1
            count_t[t[i]] = count_t.get(t[i], 0) + 1
        
        # Order both dicts and check equality
        if sorted(count_s.items()) == sorted(count_t.items()):
            return True
        return False

# Solution 3.

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        return Counter(s) == Counter(t)

# Solution 4.

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        return sorted(s) == sorted(t)

File: java/ValidAnagram.java
// Solution 1. A less optimal naive implementation that shows steps
// Time: o(n)
// Memory: o(n)^2

class ValidAnagram {
    public boolean isAnagram(String s, String t) {
        if(s.length() != t.length()) {
            return false;
        }
        
        HashMap<String, Integer> countS = new HashMap<String, Integer>();
        HashMap<String, Integer> countT = new HashMap<String, Integer>();
        
        for(int i=0; i<s.length(); i++) {
            // Get character at position i
            char char2 = s.charAt(i);
            // Convert char to string
            String c2 = String.valueOf(char2);
            // Java 8 Merge Method
            countS.merge(c2, 1, Integer::sum);  
        }
        
        for(int j=0; j<t.length(); j++) {
            // Get character at position j
            char char3 = t.charAt(j);
            // Convert char to string
            String c3 = String.valueOf(char3);
            // Java 8 Merge Method
            countT.merge(c3, 1, Integer::sum);  
        }
        
    return(countS.equals(countT));
        
    }
}